src/index.rs

```diff
@@ -179,56 +181,59 @@ impl Index {
+            self.is_ready = true;
+            return Ok(true); // no more blocks to index (done for now)
+        }
+        info!(
+            "indexing {} blocks: [{}..{}]",
+            new_headers.len(),
+            new_headers.first().unwrap().height(),
+            new_headers.last().unwrap().height()
+        );
+        for chunk in new_headers.chunks(self.batch_size) {
+            exit_flag.poll().with_context(|| {
+                format!(
+                    "indexing interrupted at height: {}",
+                    chunk.first().unwrap().height()
+                )
+            })?;
+            let blockhashes: Vec<BlockHash> = chunk.iter().map(|h| h.hash()).collect();
+            let mut heights = chunk.iter().map(|h| h.height());
 
-                let mut batch = WriteBatch::default();
-                daemon.for_blocks(blockhashes, |_blockhash, block| {
-                    let height = heights.next().expect("unexpected block");
-                    self.stats.observe_duration("block", || {
-                        index_single_block(block, height).extend(&mut batch)
-                    });
-                    self.stats.height.set("tip", height as f64);
-                })?;
-                let heights: Vec<_> = heights.collect();
-                assert!(
-                    heights.is_empty(),
-                    "some blocks were not indexed: {:?}",
-                    heights
-                );
-                batch.sort();
-                self.stats.observe_batch(&batch);
-                self.stats
-                    .observe_duration("write", || self.store.write(&batch));
-                self.stats.observe_db(&self.store);
-            }
-            self.chain.update(new_headers);
-            self.stats.observe_chain(&self.chain);
+            let mut batch = WriteBatch::default();
+            daemon.for_blocks(blockhashes, |_blockhash, block| {
+                let height = heights.next().expect("unexpected block");
+                self.stats.observe_duration("block", || {
+                    index_single_block(block, height).extend(&mut batch)
+                });
+                self.stats.height.set("tip", height as f64);
+            })?;
+            let heights: Vec<_> = heights.collect();
+            assert!(
+                heights.is_empty(),
+                "some blocks were not indexed: {:?}",
+                heights
+            );
+            batch.sort();
+            self.stats.observe_batch(&batch);
+            self.stats
+                .observe_duration("write", || self.store.write(&batch));
+            self.stats.observe_db(&self.store);
         }
-        self.store.flush();
-        Ok(())
+        self.chain.update(new_headers);
+        self.stats.observe_chain(&self.chain);
+        Ok(false) // sync is not done
+    }
+
+    pub(crate) fn is_ready(&self) -> bool {
+        self.is_ready
```

Comment:
Don't have time to properly review this huge chunk

Reply:
It should be easier to review with `--ignore-all-space`:
https://github.com/romanz/electrs/pull/543/files/24a266ea33e6dda9eed7544fb72de2e0ec2046b1?diff=split&w=1

Reply:
YIL. Was too tired yesterday, reviewed now, it looks fine.